src/xmpp/iq.c

```diff
@@ -267,8 +268,56 @@ iq_handlers_init(void)
     rooms_cache = g_hash_table_new_full(g_str_hash, g_str_equal, free, (GDestroyNotify)xmpp_stanza_release);
 }
 
+struct iq_win_finder
+{
+    gsize max, cur;
+    char** to_be_removed;
+};
+
+static void
+_win_find(char* key,
+          ProfIqHandler* handler,
+          struct iq_win_finder* finder)
+{
+    if (handler->func == _mam_rsm_id_handler) {
+        if (finder->cur >= finder->max) {
+            finder->max *= 2;
+            finder->to_be_removed = g_realloc_n(finder->to_be_removed, finder->max, sizeof(char*));
+        }
+        finder->to_be_removed[finder->cur++] = g_strdup(key);
+    }
+}
+
+void
+iq_handlers_remove_win(ProfWin* window)
+{
+    log_debug("Remove window %p of type %d", window, window ? window->type : -1);
+    if (!window)
+        return;
+    GSList *cur = late_delivery_windows, *next;
+    while (cur) {
+        LateDeliveryUserdata* del_data = cur->data;
+        next = g_slist_next(cur);
+        if (del_data->win == (void*)window)
+            late_delivery_windows = g_slist_delete_link(late_delivery_windows,
+                                                        cur);
+        cur = next;
+    }
+    struct iq_win_finder st = { 0 };
```

Comment:
Also another suggestion: make a blank space between `}` and `struct`, since these are 2 different logical parts (1 part to remove window, other to remove iq handlers, maybe these should even be separated on 2 functions).

Reply:
Generally, it's a bit strange approach to make your own dynamic array and structure just for this, why not something like this for readability? (let's call it `pseudocode` since it's not tested)

```c
static void
_mam_win_match_add(char* key, ProfIqHandler* handler, GArray* to_be_removed)
{
    if (handler->func == _mam_rsm_id_handler) {
        // Append the element to the array
        g_array_append_val(to_be_removed, g_strdup(key));
    }
}


    GArray* to_be_removed = g_array_new(FALSE, FALSE, sizeof(char*));

    // Iterate through id_handlers and call _mam_win_match_add for each entry to identify windows to be removed.
    g_hash_table_foreach(id_handlers, (GHFunc)_mam_win_match_add, to_be_removed);

    // Access elements using g_array_index and free each element
    for (guint n = 0; n < to_be_removed->len; ++n) {
        char* to_remove = g_array_index(to_be_removed, char*, n);
        g_hash_table_remove(id_handlers, to_remove);
    }
    g_array_free(to_be_removed, TRUE);
```